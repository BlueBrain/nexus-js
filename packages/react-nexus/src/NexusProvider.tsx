import * as React from 'react';
import NexusContext from './nexusContext';
import { NexusClient } from '@bbp/nexus-sdk';

export type NexusProviderProps = {
  nexusClient: NexusClient;
};

const NexusProvider: React.FC<NexusProviderProps> = ({
  nexusClient,
  children,
}) => {
  return (
    <NexusContext.Provider value={nexusClient}>
      {children}
    </NexusContext.Provider>
  );
};

export default NexusProvider;
