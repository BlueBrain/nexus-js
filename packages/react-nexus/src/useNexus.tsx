import * as React from 'react';
import invariant from 'ts-invariant';
import { NexusClient } from '@bbp/nexus-sdk';
import nexusContext from './nexusContext';

const warningMessage =
  'No Nexus client found. ' +
  'To use react-nexus components, make sure you wrap your React app with the NexusProvider component' +
  ' like: <NexusProvider nexusClient={myClient)><App /></NexusProvider>. ';

export default function useNexus<T = any, S = any>(
  apiCall: (nexus: NexusClient) => Promise<any>,
  inputs: any[] = [],
) {
  const nexus = React.useContext<NexusClient>(nexusContext);
  const [state, setState] = React.useState<{
    loading: boolean;
    error: S;
    data: T;
  }>({
    loading: true,
    error: null,
    data: null,
  });
  invariant(nexus, warningMessage);
  React.useEffect(() => {
    setState({ ...state, loading: true });
    apiCall(nexus)
      .then((data: T) => setState({ ...state, data, loading: false }))
      .catch((error: S) => setState({ ...state, error, loading: false }));
  }, inputs);

  return state;
}
