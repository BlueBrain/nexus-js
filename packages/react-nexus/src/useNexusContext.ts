import * as React from 'react';
import { NexusClient } from '@bbp/nexus-sdk';
import invariant from 'ts-invariant';
import nexusContext from './nexusContext';

const warningMessage =
  'No Nexus client found. ' +
  'To use react-nexus components, make sure you wrap your React app with the NexusProvider component' +
  ' like: <NexusProvider nexusClient={myClient)><App /></NexusProvider>. ';

export default function useNexusContext(): NexusClient {
  const nexus = React.useContext<NexusClient>(nexusContext);
  invariant(nexus, warningMessage);
  return nexus;
}
