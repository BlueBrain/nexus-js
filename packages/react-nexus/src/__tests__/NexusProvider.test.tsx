import React from 'react';
import { shallow } from 'enzyme';
import NexusProvider from '../NexusProvider'; 
import useNexusContext from '../useNexusContext';
import  useNexus  from '../useNexus';




const App = ()  => {
    const nexus = useNexusContext();
    return (
        <div>
            {nexus}
        </div>
    );
}


const AppNexus: React.SFC<any> = () => {
    const org = useNexus(nexus => nexus.Organization.get('test'), ['test']);    
    return (
        <div>
            {org}
        </div>
    );
}

describe('NexusProvider and useNexusContext',() => {
    it('useNexusContext can access the value provided by NexusProvider',  () => {
        const nexus = {
            aValue: "This is a value"
        };
    
        const result =  shallow(
            <NexusProvider nexusClient={nexus} >
                <App />
            </NexusProvider>
        );
        expect(result).toMatchSnapshot();
    });

    it('useNexus access the NexusContext through useNexusContext',  () => {
        const nexus = {
            Organization: {
                list: [{ name: 'org1'}]
            }
        };
          
        const result =  shallow(
            <NexusProvider nexusClient={nexus} >
                <AppNexus />
            </NexusProvider>
        );
        expect(result).toMatchSnapshot();      
        
    })    
})
