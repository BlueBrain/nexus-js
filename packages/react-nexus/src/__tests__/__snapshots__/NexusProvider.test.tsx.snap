// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NexusProvider and useNexusContext useNexus access the NexusContext through useNexusContext 1`] = `
<ContextProvider
  value={
    Object {
      "Organization": Object {
        "list": Array [
          Object {
            "name": "org1",
          },
        ],
      },
    }
  }
>
  <AppNexus />
</ContextProvider>
`;

exports[`NexusProvider and useNexusContext useNexusContext can access the value provided by NexusProvider 1`] = `
<ContextProvider
  value={
    Object {
      "aValue": "This is a value",
    }
  }
>
  <App />
</ContextProvider>
`;
